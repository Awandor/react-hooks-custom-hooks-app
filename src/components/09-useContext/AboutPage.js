import { useContext } from "react";
import { UserContext } from "./UserContext";

export const AboutPage = () => {

    const { user, setUser } = useContext( UserContext );

    const handleClick = () => {

        setUser( {} );

    };

    return (
        <div className="container">
            <h1>About Page</h1>
            <hr />

            <pre>{JSON.stringify( user, null, 3 )}</pre>

            <button className="btn btn-warning" onClick={handleClick}>Log out</button>

        </div>
    );

};
