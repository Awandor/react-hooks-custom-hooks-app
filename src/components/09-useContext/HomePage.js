import { useContext } from "react";
import { UserContext } from "./UserContext";

export const HomePage = () => {

    // const userContext = useContext( UserContext );

    // console.log( userContext );

    const { user } = useContext( UserContext );

    

    return (
        <div className="container">
            <h1>Home Page</h1>
            <hr />

            <pre>
                {JSON.stringify( user, null, 3 )}
            </pre>
            
        </div>
    );

};
