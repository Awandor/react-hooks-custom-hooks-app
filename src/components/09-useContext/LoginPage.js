import { useContext } from "react";
import { UserContext } from "./UserContext";

export const LoginPage = () => {

    const { setUser } = useContext( UserContext );

    const algo = {
        id:3333,
        name: 'Anders'
    };

    return (
        <div className="container">
            <h1>Login Page</h1>
            <hr/>

            <button className="btn btn-primary" onClick={()=> setUser( algo )}>Login</button>
            
        </div>
    );

};
