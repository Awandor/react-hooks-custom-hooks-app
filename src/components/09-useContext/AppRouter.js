import { BrowserRouter as Router, Switch, Route, Redirect } from 'react-router-dom';
import { AboutPage } from './AboutPage';
import { HomePage } from './HomePage';
import { LoginPage } from './LoginPage';
import { NavBar } from './NavBar';

export const AppRouter = () => {

    return (
        <Router>
            <div>
                <NavBar />
                
                <Switch>
                    <Route exact path="/about" component={AboutPage} />

                    <Route exact path="/login" component={LoginPage} />

                    {/* Se tiene que poner en último lugar de lo contrario se irá ahí siempre */}

                    <Route exact path="/" component={HomePage} />

                    {/* Aunque se puede poner el atributo exact y sólo se cumple si es exactamente y ya se podría poner en otro orden */}

                    <Route component={HomePage} />

                    {/* Funciona como el default de switch */}

                    {/* Otra manera de hacerlo es con Redirect */}

                    <Redirect to="/" />

                </Switch>
            </div>
        </Router>
    );

};
