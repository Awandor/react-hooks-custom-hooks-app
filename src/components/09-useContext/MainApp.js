import { useState } from "react";
import { AppRouter } from "./AppRouter"
import { UserContext } from "./UserContext";

export const MainApp = () => {

    /* const user = {
        id: 1234,
        nombre: 'Dan',
        email: 'algo@algo.com'
    } */

    const [ user, setUser ] = useState( {} );

    /* return (
        <UserContext.Provider value={user}>
            <AppRouter />
        </UserContext.Provider>
    ); */

    // Pasamos un objeto como value con el m√©todo setUser

    /* return (
        <UserContext.Provider value={{user: user, setUser: setUser}}>
            <AppRouter />
        </UserContext.Provider>
    ); */

    return (
        <UserContext.Provider value={{ user, setUser }}>
            <AppRouter />
        </UserContext.Provider>
    );

};
