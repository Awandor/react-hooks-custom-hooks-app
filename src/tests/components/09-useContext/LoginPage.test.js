import { mount } from 'enzyme';
import { LoginPage } from '../../../components/09-useContext/LoginPage';
import { UserContext } from '../../../components/09-useContext/UserContext';

describe( 'Prueba en <LoginPage />', () => {

    // Tenemos que crear un context de lo contrario falla

    const setUser = jest.fn();

    const wrapper = mount( <UserContext.Provider value={ { setUser: setUser } }> <LoginPage /> </UserContext.Provider> );

    
    
    test( 'debe mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();
        
    } );

    test( 'debe ejecutarse setUser con el argumento esperado', () => {

        const algo = {
            id:3333,
            name: 'Anders'
        };

        wrapper.find( 'button' ).simulate( 'click' );

        console.log( 'button', wrapper.find( 'button' ) );

        expect( setUser ).toHaveBeenCalledWith( algo );
        
    } );
    
    
} );
