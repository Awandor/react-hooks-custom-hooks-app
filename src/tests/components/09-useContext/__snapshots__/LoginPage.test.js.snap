// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <LoginPage /> debe mostrarse correctamente 1`] = `
Array [
  " ",
  <div
    className="container"
  >
    <h1>
      Login Page
    </h1>
    <hr />
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Login
    </button>
  </div>,
  " ",
]
`;
